### ISSUE TITLE
issue with type narrowing on literal True

### ISSUE THREAD

**Issue by Daggy1234** [microsoft/pyright#2056]
**Describe the bug**
Issue with type narrowing for literals. When Using a Union with literals.

**To Reproduce**
Checking code below with SublimeText4 lsp-pyright, use code below. Pyright will error.

**Expected behavior**
It should narrow type dude to the first condition of union, and predict it is a TypedDict and not a str.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from typing import Literal, Tuple, TypedDict, Union


class OutDict(TypedDict):
    message: str


SUCCESS_TYPE = Tuple[Literal[True], OutDict]
ERROR_TYPE = Tuple[Literal[False], str]

def fn(sucess: bool) -> Union[SUCCESS_TYPE, ERROR_TYPE]:
    if sucess:
        return True, {"message": "works!"}
    return False, "No way"



status, data = fn(False)

if status:
    print(data["message"])
```

![Error](https://media.discordapp.net/attachments/336642776609456130/860919302144131103/unknown.png?width=1921&height=231)

```
   21:8   error   Argument of type "Literal['message']" cannot be assigned to parameter "i" of type "int | slice" in function "__getitem__" ​Pyright:
                   Type "Literal['message']" cannot be assigned to type "int | slice"
                     "Literal['message']" is incompatible with "int"
                     "Literal['message']" is incompatible with "slice
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.
```shell
pyright 1.1.154
Python 3.9.5
```
**Additional context**

**Comment by erictraut** [comment]
This is expected behavior. Type associations are not maintained across different symbols. In your example above, `status` and `data` are independent symbols each with its own type.

Here are some alternative approaches that work with type checking:
```python
class SuccessType(NamedTuple):
    success: Literal[True]
    data: OutDict

class ErrorType(NamedTuple):
    success: Literal[False]
    data: str

def fn(sucess: bool) -> Union[SuccessType, ErrorType]:
    if sucess:
        return SuccessType(True, {"message": "works!"})
    return ErrorType(False, "No way")

result = fn(False)

if result.success == True:
    print(result.data["message"])
```

or

```python
ErrorResult = NewType("ErrorResult", str)

def fn(sucess: bool) -> Union[OutDict, ErrorResult]:
    if sucess:
        return {"message": "works!"}
    return ErrorResult("No way")

result = fn(False)

if isinstance(result, OutDict):
    print(result["message"])
```

**Comment by numbermaniac** [comment]
One thing that surprises me, though, is that if I take Daggy1234's code and replace the last 3 lines with the following:

```py
result = fn(False)

if result[0] == True:
    print(result[1]["message"])
```

then it still does not work - I get the same error message.

I figured that this might work since the two values are still "contained" together by remaining in the tuple. Shouldn't the if-statement, in theory, be sufficient for pyright to determine that `result` must be of type `SUCCESS_TYPE` only?

**Comment by erictraut** [comment]
That isn't a type narrowing pattern that pyright supports. For a full list of supported patterns, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-narrowing). Mypy (another Python type checker) also does not support this pattern.

Each type narrowing pattern requires separate custom logic and has the potential to slow down type analysis, so we add support for new patterns only when there's good evidence that it is a common pattern. I haven't seen this particular approach (using a tuple with literal bool types), and I suspect that it's not a common approach. We would consider adding this support if there's sufficient interest (a formal feature request filed in the pylance-release repo with a significant number of upvotes over the course of several months).
